core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java

```diff
@@ -0,0 +1,690 @@
+          return file.getName().startsWith(taskNo) && file.getName()
+              .endsWith(CarbonTablePath.INDEX_FILE_EXT);
+        }
+      });
+      if (carbonFiles != null && carbonFiles.length > 0) {
+        SegmentFile segmentFile = new SegmentFile();
+        Map<String, FolderDetails> locationMap = new HashMap<>();
+        FolderDetails folderDetails = new FolderDetails();
+        folderDetails.setRelative(isRelative);
+        folderDetails.setPartitions(partionNames);
+        folderDetails.setStatus(SegmentStatus.SUCCESS.getMessage());
+        for (CarbonFile file : carbonFiles) {
+          folderDetails.getFiles().add(file.getName());
+        }
+        locationMap.put(location, folderDetails);
+        segmentFile.setLocationMap(locationMap);
+        String path = writePath + "/" + taskNo + CarbonTablePath.SEGMENT_EXT;
+        // write segment info to new file.
+        writeSegmentFile(segmentFile, path);
+      }
+    }
+  }
+
+  /**
+   * Writes the segment file in json format
+   * @param segmentFile
+   * @param path
+   * @throws IOException
+   */
+  public void writeSegmentFile(SegmentFile segmentFile, String path) throws IOException {
+    AtomicFileOperations fileWrite =
+        new AtomicFileOperationsImpl(path, FileFactory.getFileType(path));
+    BufferedWriter brWriter = null;
+    DataOutputStream dataOutputStream = null;
+    Gson gsonObjectToWrite = new Gson();
+    try {
+      dataOutputStream = fileWrite.openForWrite(FileWriteOperation.OVERWRITE);
+      brWriter = new BufferedWriter(new OutputStreamWriter(dataOutputStream,
+          Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET)));
+
+      String metadataInstance = gsonObjectToWrite.toJson(segmentFile);
+      brWriter.write(metadataInstance);
+    } finally {
+      if (null != brWriter) {
+        brWriter.flush();
+      }
+      CarbonUtil.closeStreams(brWriter);
+      fileWrite.close();
+    }
+  }
+
+  /**
+   * Merge all segment files in a segment to single file.
+   *
+   * @param writePath
+   * @throws IOException
+   */
+  public SegmentFile mergeSegmentFiles(String readPath, String mergeFileName, String writePath)
+      throws IOException {
+    CarbonFile[] segmentFiles = getSegmentFiles(readPath);
+    if (segmentFiles != null && segmentFiles.length > 0) {
+      SegmentFile segmentFile = null;
+      for (CarbonFile file : segmentFiles) {
+        SegmentFile localSegmentFile = readSegmentFile(file.getAbsolutePath());
+        if (segmentFile == null && localSegmentFile != null) {
+          segmentFile = localSegmentFile;
+        }
+        if (localSegmentFile != null) {
+          segmentFile = segmentFile.merge(localSegmentFile);
+        }
+      }
+      if (segmentFile != null) {
+        String path = writePath + "/" + mergeFileName + CarbonTablePath.SEGMENT_EXT;
+        writeSegmentFile(segmentFile, path);
+        FileFactory.deleteAllCarbonFilesOfDir(FileFactory.getCarbonFile(readPath));
```

Comment:
Use try and finally here and call FileFactory.deleteAllCarbonFilesOfDir(FileFactory.getCarbonFile(readPath)) inside finally block

Reply:
Failure case is handled in abortJob committer

Reply:
cleanup required for temporary folders, if abortJob is not called when process is killed.